event_files/topdown.json

```diff
@@ -0,0 +1,96 @@
+    },
+    {
+        "EventName": "TOPDOWN.BAD_SPECULATION.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots wasted due to incorrect speculations.",
+        "PublicDescription": "TMA slots wasted due to incorrect speculations. This include slots used to issue uops that do not eventually get retired and slots for which the issue-pipeline was blocked due to recovery from earlier incorrect speculation. For example; wasted slots due to miss-predicted branches should be accounted by this event. Incorrect data speculation followed by Memory Ordering Nukes is another example."
+    },
+    {
+        "MetricName": "TOPDOWN.BAD_SPECULATION.SLOTS_V2",
+        "BriefDescription": "TMA slots wasted due to incorrect speculations",
+        "MetricExpr": "SPEC.UOP_ISSUED - RETIRED.UOP + pipeline_width * PRD.PIPELINE_FLUSH.RECOVERY_CYCLES"
+    },
+    {
+        "EventName": "TOPDOWN.BAD_SPECULATION.CONTROL_FLOW.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots wasted due to incorrect control flow speculations.",
+        "PublicDescription": "TMA slots wasted due to incorrect control flow speculations."
+    },
+    {
+        "EventName": "TOPDOWN.BAD_SPECULATION.MEM_ORDERING.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots wasted due to memory ordering violations.",
+        "PublicDescription": "TMA slots wasted due to memory ordering violations."
+    },
+    {
+        "EventName": "TOPDOWN.BACKEND_BOUND.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots unused due to the lack of backend resources",
+        "PublicDescription": "TMA slots unused due to the lack of backend resources"
+    },
+    {
+        "EventName": "TOPDOWN.BACKEND_BOUND.MEMORY_BOUND.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots unused due to the stalls caused by load and store instructions",
+        "PublicDescription": "TMA slots unused due to the stalls caused by load and store instructions"
+    },
+    {
+        "EventName": "TOPDOWN.BACKEND_BOUND.CORE_BOUND.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots unused due to the non-memory stalls",
+        "PublicDescription": "TMA slots unused due to the non-memory stalls"
+    },
+    {
+        "MetricName": "TOPDOWN.BACKEND_BOUND.CORE_BOUND.SLOTS_V2",
+        "BriefDescription": "TMA slots unused due to the non-memory stalls",
+        "MetricExpr": "TOPDOWN.BACKEND_BOUND.SLOTS - TOPDOWN.BACKEND_BOUND.MEMORY_BOUND.SLOTS"
+    },
+    {
+        "EventName": "TOPDOWN.BACKEND_BOUND.MEMORY_BOUND.L1_BOUND.SLOTS",
+        "EventCode": "0x0",
+        "BriefDescription": "TMA slots unused due to the stalls caused by load instructions without missing L1 data cache",
+        "PublicDescription": "TMA slots unused due to the stalls caused by load instructions without missing L1 data cache"
+    },
```

Comment:
I recommend we do something better than what Intel does for L1_bound.  L1_bound seems to be a catch-all, and includes a lot of stalls for ops that haven't even looked up the L1 yet, and may end up missing it.  What if we had these MEM_BOUND sub-events:

* TRANSLATION - slots wasted while waiting for address translation
* DATA - slots wasted waiting for data

There would still be some "other" cases here, but they are likely more microarchitectural in nature ("X buffer full").  Then these events could be broken down into:

ADDR: AGEN, L1TLB, L2TLB, (more?), MMU
DATA: STLF, L1, L2, L3

AGEN = address generation, i.e. waiting for older ops to complete to get the address for this load/store.  Admittedly I'm not sure how likely such a load/store is to be the cause of backend-bound, so we can think about how useful that one is.  STLF = store-to-load forward.  Arguably we could include DRAM in DATA_BOUND, but there are so many options now (DRAM, HBM, CXL, etc) that maybe we just leave those for other (or custom events per memory source supported). 

For ADDR-bound, these can be loads or stores.  For DATA-bound, it should probably be up to implementations.  Certainly it would include loads.  If stores are executed post-retirement then they should not be included, but, for a given implementation, some/all stores may execute -pre-retirement.

I see below that you had TLB_BOUND, so maybe you were thinking along the same lines, and I'm just suggesting a bit more breakdown.  Also, for sub-events of MEMORY_BOUND it seems like we could leave off the "_BOUND".  JMO.

Reply:
Where do stalls that result from serialization land?  

Arguably slots stalled due to FENCE/FENCE.TSO should qualify as MEMORY_BOUND, since we're stalled waiting for mem ops.  But FENCE.I is a trickier case, since we stall fetch waiting for stores to complete.  So we'd be FE_BOUND, not BE_BOUND.  Do we plan to have FE_BOUND sub-events?

Also, there is microarchitectural serialization.  E.g., writes to some CSRs serialize to ensure that younger insts observe the change.  These likely fall under CORE_BOUND.

Reply:
1. Yes let's try applying this hierarchy. I like the split between ADDR and DATA. Although regarding the ADDR - not sure if it should contain AGEN part because basically it means those older uops are real bottleneck. And what kinds of perf bottlenecks can fall under MMU?
2. Serialization instructions is an overlook it seems, need to add it. Intel has it under Core Bound as I see.